src/java.base/share/classes/java/lang/Thread.java

```diff
@@ -372,81 +554,497 @@ public static void sleep(long millis, int nanos)
     public static void onSpinWait() {}
 
     /**
-     * Initializes a Thread.
+     * Characteristic value signifying that the thread cannot set values for its
+     * copy of {@link ThreadLocal thread-locals}.
+     * See Thread initialization.
+     */
+    static final int NO_THREAD_LOCALS = 1 << 1;
+
+    /**
+     * Characteristic value signifying that initial values for {@link
+     * InheritableThreadLocal inheritable-thread-locals} are not inherited from
+     * the constructing thread.
+     * See Thread initialization.
+     */
+    static final int NO_INHERIT_THREAD_LOCALS = 1 << 2;
+
+    /**
+     * Helper class to generate thread identifiers. The identifiers start at
+     * 2 as this class cannot be used during early startup to generate the
+     * identifier for the primordial thread. The counter is off-heap and
+     * shared with the VM to allow it assign thread identifiers to non-Java
+     * threads.
+     * See Thread initialization.
+     */
+    private static class ThreadIdentifiers {
+        private static final Unsafe U;
+        private static final long NEXT_TID_OFFSET;
+        static {
+            U = Unsafe.getUnsafe();
+            NEXT_TID_OFFSET = Thread.getNextThreadIdOffset();
+        }
+        static long next() {
+            return U.getAndAddLong(null, NEXT_TID_OFFSET, 1);
+        }
+    }
+
+    /**
+     * Returns the context class loader to inherit from the parent thread.
+     * See Thread initialization.
+     */
+    private static ClassLoader contextClassLoader(Thread parent) {
+        @SuppressWarnings("removal")
+        SecurityManager sm = System.getSecurityManager();
+        if (sm == null || isCCLOverridden(parent.getClass())) {
+            return parent.getContextClassLoader();
+        } else {
+            // getContextClassLoader not trusted
+            ClassLoader cl = parent.contextClassLoader;
```

Comment:
I might be misreading the comment on that line, but reading this if/else block a few times, I'm unsure what the expectations here are.

It's my understanding that a call to `getContextClassLoader()` can't be trusted if that method has been overridden by the passed `Thread` type. In such cases, we don't call that method and instead just use the field value of `contextClassLoader` (which is a private field on `Thread`). Is that understanding correct? If yes, then the condition in the `if` block a few lines above, looks odd. It seems to be calling the `getContextClassLoader()` if  that method is overridden by the passed `Thread` type, i.e. the untrusted case. Should it instead be:
```
if (sm == null || !isCCLOverridden(parent.getClass())) {
    return parent.getContextClassLoader();
}
```
(notice the negation)

Reply:
This code hasn't changed, it's just moved to a helper method. When running with a security manager and the CCL methods aren't overridden, then it avoids the permission check. However, I can see how the comment is mis-reading so we can improve that.